nupic/research/frameworks/vernon/mixins/representation_overlap.py

```diff
@@ -0,0 +1,177 @@
+
+__all__ = [
+    "PlotRepresentationOverlap",
+]
+
+
+class PlotRepresentationOverlap(metaclass=abc.ABCMeta):
+    """
+    Mixin for plotting a module's inter- and intra-class representation overlap.
+
+    :param config: a dict containing the following
+
+        - plot_representation_overlap_args: a dict containing the following
+
+            - include_modules: (optional) a list of module types to track
+            - include_names: (optional) a list of module names to track e.g.
+                             "features.stem"
+            - include_patterns: (optional) a list of regex patterns to compare to the
+                                names; for instance, all feature parameters in ResNet
+                                can be included through "features.*"
+            - plot_freq: (optional) how often to create the plot, measured in training
+                         iterations; defaults to 1
+            - plot_args: (optional) either a dictionary or a callable that takes no
+                             arguments and returns a dictionary; for instance this may
+                             be used to return a random sample of integers specifying
+                             units to plot; called only once at setup
+            - max_samples_to_plot: (optional) how many of samples to use for plotting;
+                                   only the newest will be used; defaults to 5000
+
+    Example config:
+    ```
+    config=dict(
+        plot_representation_overlap_args=dict(
+            include_modules=[torch.nn.ReLU, KWinners],
+            plot_freq=1,
+            plot_args=dict(annotate=False),
+            max_samples_to_plot=2000
+        )
+    )
+    ```
+    """
+
+    def setup_experiment(self, config):
+        super().setup_experiment(config)
+
+        # Process config args
+        ro_args = config.get("plot_representation_overlap_args", {})
+        ro_plot_freq, filter_args, ro_max_samples = self.process_ro_args(ro_args)
+
+        self.ro_plot_freq = ro_plot_freq
+        self.ro_max_samples = ro_max_samples
+
+        # Register hook for tracking hidden activations - useful for representation
+        # overlap
+        named_modules = filter_modules(self.model, **filter_args)
+        hook_args = dict(max_samples_to_track=self.ro_max_samples)
+        self.ro_hook = ModelHookManager(named_modules,
+                                        TrackHiddenActivationsHook,
+                                        hook_args=hook_args)
+
+        # Log the names of the modules being tracked
+        tracked_names = pformat(list(named_modules.keys()))
+        self.logger.info(f"Tracking representation overlap of modules: {tracked_names}")
+
+        # The targets will be collected in `self.error_loss` in a 1:1 fashion
+        # to the tensors being collected by the hooks.
+        self.ro_targets = torch.tensor([]).long()
+
+    def process_ro_args(self, ro_args):
```

Comment:
One thing you can do, so you don't have to distinguish the name of `process_args` with the `ro`, is to define `process_ro_args` as a stand-alone function in the same file. Which you can do since it doesn't require `self` anywhere in the function. This way, you won't have to worry about any conflicts with inheritance. Tbh, I should have done this with the mixins I wrote too.

Reply:
I agree; it's now a separate function.